rfcs/0135-custom-asserts.md

```diff
@@ -0,0 +1,116 @@
+---
+feature: custom-asserts
+start-date: 2022-09-21
+author: Anselm Schüler
+co-authors: None
+shepherd-team: None
+shepherd-leader: None
+related-issues: None
+---
+
+# Summary
+[summary]: #summary
+
+Allow users to use attribute sets with a boolean attribute `success` and a string attribute `message` instead of a boolean in `assert …; …` expressions.
+
+# Motivation
+[motivation]: #motivation
+
+Since Nix is an untyped language, asserts are often needed to ensure that a function or program works correctly. However, the current assert system is unsuitable for more sophisticated error reporting that aims to inform the user as to what has happened.
+
+Consider a nixpkgs package expression that wants to validate its arguments. Currently, the best way to
+provide a custom error message is to use `assert … -> throw …; …`.
+This method has several disadvantages: Since an implication from a falsehood is always true, you are required
+to invert the condition. Additionally, since the assertion itself is not triggered by the error,
```

Comment:
Missed this bit on my first read through. Note that this drawback is not an issue if you do `assert condition || throw "…"; …`. No need to invert, quite straightforward.

Reply:
OK, I removed that.